pandas/tests/indexes/test_multi.py

```diff
@@ -1317,6 +1317,18 @@ def test_get_indexer_nearest(self):
         with tm.assertRaises(NotImplementedError):
             midx.get_indexer(['a'], method='pad', tolerance=2)
 
+    def test_hash_collisions(self):
+        # non-smoke test that we don't get hash collisions
+
+        index = MultiIndex.from_product([np.arange(1000), np.arange(1000)],
+                                        names=['one','two'])
+        result = index.get_indexer(index.values)
+        self.assert_numpy_array_equal(result, np.arange(len(index), dtype='int64'))
+
+        for i in [0, 1, len(index)-2, len(index)-1]:
+            result = index.get_loc(index[i])
+            self.assertEqual(result, i)
```

Comment:
The only fool-proof way to prevent mistaken lookups is to compare the tuples for equality -- keeping around big arrays of tuples is how we got into this mess in the first place. So you can

* compute the hashed key value
* look up the index location
* construct the tuple for that index location
* compare the tuple in the index with your tuple key
* if they are unequal, raise a KeyError

With `reindex` operations, it gets more expense. But at least we aren't keeping around an huge internal array of tuples like we were before.